apps/aecore/src/aec_conductor.erl

```diff
@@ -582,101 +620,15 @@ handle_mining_reply({{error, {runtime, Reason}}, _}, State) ->
 
-retry_mining(#state{block_candidate = #candidate{top_hash = Hash1},
-                    seen_top_block_hash = Hash2} = State) when Hash1 =/= Hash2 ->
-    %% Stale hash for candidate. Rebuild it.
-    epoch_mining:info("Stale hash for candidate"),
-    start_mining(State#state{block_candidate = undefined});
-retry_mining(#state{block_candidate = Candidate,
-                    fetch_new_txs_from_pool = true} = State) ->
-    Block = Candidate#candidate.block,
-    %% We should first see if we can get a new candidate.
-    case aec_mining:need_to_regenerate(Block) of
-        false ->
-            epoch_mining:debug("No new txs available"),
-            retry_mining_with_new_nonce(Candidate, State);
-        true ->
-            epoch_mining:debug("New txs added for mining"),
-            start_mining(State#state{block_candidate = undefined})
-    end;
-retry_mining(#state{block_candidate = Candidate} = State) ->
-    retry_mining_with_new_nonce(Candidate, State).
+retry_mining(S = #state{ block_candidate = Candidate }) ->
+    start_mining(S#state{ block_candidate = bump_nonce(Candidate) }).
 
-retry_mining_with_new_nonce(#candidate{nonce = N, max_nonce = N}, State) ->
+bump_nonce(#candidate{ nonce = N, max_nonce = N }) ->
     epoch_mining:info("Failed to mine block, "
                       "nonce range exhausted (was ~p)", [N]),
-    start_mining(State#state{block_candidate = undefined});
-retry_mining_with_new_nonce(Candidate, State) ->
-    Nonce = Candidate#candidate.nonce,
-    NewNonce = aec_pow:next_nonce(Nonce),
-    epoch_mining:debug("Not fetching new txs; "
-                       "continuing mining with bumped nonce "
-                       "(was ~p, is ~p).", [Nonce, NewNonce]),
-    NewCandidate = Candidate#candidate{nonce = NewNonce},
-    start_mining(State#state{block_candidate = NewCandidate}).
-
-%%%===================================================================
-%%% Worker: Generate new block candidates
-
-new_candidate(Block, Nonce, MaxNonce, State) ->
-    HeaderBin = aec_headers:serialize_to_binary(aec_blocks:to_header(Block)),
-    #candidate{block = Block,
-               bin = HeaderBin,
-               nonce = Nonce,
-               max_nonce = MaxNonce,
-               top_hash = State#state.seen_top_block_hash
-              }.
-
-create_block_candidate(#state{keys_ready = false} = State) ->
-    %% Keys are needed for creating a candidate
-    wait_for_keys(State);
-create_block_candidate(State) ->
-    epoch_mining:info("Creating block candidate"),
-    Fun = fun() ->
-                  {TopBlock, TopBlockState} = aec_chain:top_block_with_state(),
-                  AdjChain = get_adjustment_headers(TopBlock),
-                  {aec_mining:create_block_candidate(TopBlock, TopBlockState,
-                                                     AdjChain),
-                   State#state.seen_top_block_hash}
-          end,
-    dispatch_worker(create_block_candidate, Fun, State).
```

Comment:
Documentation at top of module seems to assume that creation of block candidate is still worker - may benefit from rephrasing though optional.
https://github.com/aeternity/epoch/blob/76053d4e6149078f20f6a129ca119d4a62b0ecdc/apps/aecore/src/aec_conductor.erl#L10-L11

Reply:
Rephrased.